#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [2]:
import pandas as pd

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
data_path = ('/content/drive/MyDrive/Data Analyst/Python/Data/supermarket_sales.csv')
df = pd.read_csv(data_path)

In [ ]:
df.head(10)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3
5,699-14-3026,C,Naypyitaw,Normal,Male,Electronic accessories,85.39,7,29.8865,627.6165,3/25/2019,18:30,Ewallet,597.73,4.761905,29.8865,4.1
6,355-53-5943,A,Yangon,Member,Female,Electronic accessories,68.84,6,20.6520,433.6920,2/25/2019,14:36,Ewallet,413.04,4.761905,20.6520,5.8
7,315-22-5665,C,Naypyitaw,Normal,Female,Home and lifestyle,73.56,10,36.7800,772.3800,2/24/2019,11:38,Ewallet,735.60,4.761905,36.7800,8.0
8,665-32-9167,A,Yangon,Member,Female,Health and beauty,36.26,2,3.6260,76.1460,1/10/2019,17:15,Credit card,72.52,4.761905,3.6260,7.2
9,692-92-5582,B,Mandalay,Member,Female,Food and beverages,54.84,3,8.2260,172.7460,2/20/2019,13:27,Credit card,164.52,4.761905,8.2260,5.9


1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [ ]:
total_profit = df.groupby('Branch')['gross income'].sum().sort_values(ascending=False).round(3)
total_profit

,gross income
Branch,
C,5265.176
A,5057.160
B,5057.032


In [ ]:
total_profit = df.groupby('Branch')['gross income'].sum()
total_profit_pct = (total_profit / total_profit.sum() * 100).round(4)

print("Сумарний прибуток:")
print(total_profit.round(3))
print("\nЧастка від загального прибутку (%):")
print(total_profit_pct)

Сумарний прибуток:
Branch
A    5057.160
B    5057.032
C    5265.176
Name: gross income, dtype: float64

Частка від загального прибутку (%):
Branch
A    32.8828
B    32.8819
C    34.2353
Name: gross income, dtype: float64


In [ ]:
total_profit = df.groupby('Branch')['gross income'].sum().sort_values(ascending=False)
most_profitable = total_profit.idxmax()
max_profit = total_profit.max()

print(f"Найприбутковіший філіал: {most_profitable}")
print(f"Прибуток: {max_profit:.2f}")

Найприбутковіший філіал: C
Прибуток: 5265.18


2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [ ]:
city_branch_profit = df.groupby(['City', 'Branch'])['gross income'].sum().sort_values(ascending=False).round(3)
city_branch_profit

,,gross income
City,Branch,
Naypyitaw,C,5265.176
Yangon,A,5057.160
Mandalay,B,5057.032


In [ ]:
city_branch_profit = df.groupby(['City', 'Branch'])['gross income'].sum()
top_branch = city_branch_profit.idxmax()

print(f"Найприбутковіший філіал: {top_branch[1]}")
print(f"Місто: {top_branch[0]}")
print(f"Прибуток: {city_branch_profit.max():.2f}")

Найприбутковіший філіал: C
Місто: Naypyitaw
Прибуток: 5265.18


3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [ ]:
invoices_by_category = pd.pivot_table(
    df,
    index='Branch',
    columns='Product line',
    values='Invoice ID',
    aggfunc='count'
)
invoices_by_category

Product line,Electronic accessories,Fashion accessories,Food and beverages,Health and beauty,Home and lifestyle,Sports and travel
Branch,,,,,,
A,60,51,58,47,65,59
B,55,62,50,53,50,62
C,55,65,66,52,45,45


In [ ]:
invoices_by_category = pd.crosstab(
    df['Branch'],
    df['Product line']
)
invoices_by_category

Product line,Electronic accessories,Fashion accessories,Food and beverages,Health and beauty,Home and lifestyle,Sports and travel
Branch,,,,,,
A,60,51,58,47,65,59
B,55,62,50,53,50,62
C,55,65,66,52,45,45


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [ ]:
electronic_column = invoices_by_category['Electronic accessories'].sort_values(ascending=False)
print(electronic_column)
print('\n')
best_branch = electronic_column.idxmax()
max_values = electronic_column.max()

print(f'Філіал з найбільшою кількістю: {best_branch} = {max_values}')

Branch
A    60
B    55
C    55
Name: Electronic accessories, dtype: int64


Філіал з найбільшою кількістю: A = 60


4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


In [ ]:
df.head(10)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3
5,699-14-3026,C,Naypyitaw,Normal,Male,Electronic accessories,85.39,7,29.8865,627.6165,3/25/2019,18:30,Ewallet,597.73,4.761905,29.8865,4.1
6,355-53-5943,A,Yangon,Member,Female,Electronic accessories,68.84,6,20.6520,433.6920,2/25/2019,14:36,Ewallet,413.04,4.761905,20.6520,5.8
7,315-22-5665,C,Naypyitaw,Normal,Female,Home and lifestyle,73.56,10,36.7800,772.3800,2/24/2019,11:38,Ewallet,735.60,4.761905,36.7800,8.0
8,665-32-9167,A,Yangon,Member,Female,Health and beauty,36.26,2,3.6260,76.1460,1/10/2019,17:15,Credit card,72.52,4.761905,3.6260,7.2
9,692-92-5582,B,Mandalay,Member,Female,Food and beverages,54.84,3,8.2260,172.7460,2/20/2019,13:27,Credit card,164.52,4.761905,8.2260,5.9


###Гіпотеза №1
Чи правда, що покупці, які залишають високий рейтинг, витрачають більше грошей, ніж ті, хто ставить низький рейтинг?

In [9]:
def rating_mapping(val):
  if val >= 8:
    return 'High'
  elif 6 <= val < 8:
    return 'Medium'
  else:
    return 'Low'

df['Rating Category'] = df.Rating.apply(rating_mapping)


In [12]:
df.head(5)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating,Rating Category
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1,High
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6,High
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4,Medium
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4,High
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3,Low


In [10]:
df.groupby('Rating Category')['Total'].mean().sort_values(ascending=False).round(3)

,Total
Rating Category,
Low,340.572
High,323.281
Medium,307.098


Парадокс:
Незадоволені клієнти (низький рейтинг) витрачають БІЛЬШЕ грошей, ніж задоволені!
Це протилежно тому, що очікувалося отримати!

In [11]:
# Перевірка скільки людей в кожній категорії:
df.groupby('Rating Category').size().sort_values(ascending=False)

,0
Rating Category,
Medium,356
High,329
Low,315


Ключовий інсайт:
Близько третини клієнтів (315 людей) залишають низький рейтинг, але витрачають на 10% більше (340.57 vs 307.10), ніж задоволені клієнти. Це означає, що найцінніші клієнти незадоволені сервісом — критична проблема для бізнесу.

Бізнес-ризик:
Якщо ці 315 клієнтів не повернуться через незадоволеність, супермаркет втратить ~107,000 доходу (315 × 340.57), що становить понад 30% від загального обороту.

Рекомендація:
Терміново покращити досвід для клієнтів з великими чеками (>300) — швидші каси, VIP-зони, допомога з паковкою. Це утримає найприбутковіших клієнтів.

###Гіпотеза №2
Чи правда, що в обідній час (13:00-15:00) купують більше Food and beverages, ніж в інші години?

In [17]:
def get_hour(time):
    return int(time.split(':')[0])

df['Hour'] = df.Time.apply(get_hour)
df.head()

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating,Rating Category,Hour
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1,High,13
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6,High,10
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4,Medium,13
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4,High,20
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3,Low,10


In [18]:
# Обідній час
lunch_food = df[(df['Hour'].isin([13, 14, 15])) & (df['Product line'] == 'Food and beverages')]

# Решта часу
non_lunch_food = df[(~df['Hour'].isin([13, 14, 15])) & (df['Product line'] == 'Food and beverages')]

print(f"Обід (13-15): {len(lunch_food)} інвойсів")
print(f"Не-обід: {len(non_lunch_food)} інвойсів")

Обід (13-15): 48 інвойсів
Не-обід: 126 інвойсів


In [19]:
# Які категорії популярні в обід?
lunch_categories = df[df['Hour'].isin([13, 14, 15])]['Product line'].value_counts()
print(lunch_categories)

Product line
Fashion accessories       52
Sports and travel         50
Home and lifestyle        48
Health and beauty         48
Food and beverages        48
Electronic accessories    42
Name: count, dtype: int64


In [20]:
# Кількість Food and beverages по кожній годині
food_by_hour = df[df['Product line'] == 'Food and beverages'].groupby('Hour').size().sort_values(ascending=False)
print(food_by_hour)

Hour
19    26
15    21
10    20
20    20
18    17
16    16
13    15
14    12
11    11
12    10
17     6
dtype: int64


###Ключовий інсайт
####Піки покупок їжі:

######19:00 (26) — вечір, після роботи, покупки на вечерю
######15:00 (21) — кінець обідньої перерви, підготовка до вечора
######10:00, 20:00 (по 20) — ранок та пізній вечір

####Провали:

######17:00 (6) — люди ще на роботі або в дорозі додому
######13:00, 14:00 (15, 12) — обідня перерва, але їдять в кафе/їдальнях, не купують в супермаркеті

---

###Бізнес-рекомендації
####Не працює:
#####Обідні lunch deals та промо на їжу в супермаркеті (13-15 год) — люди обідають в інших місцях.

####Працює:
#####Вечірній фокус (19:00): Готові вечері, сімейні набори, fresh продукти — максимальний попит
#####Післяобідній час (15:00): Снеки, готова їжа "додому", напої
Ранок (10:00): Продукти на день, готові сніданки

####Дія: Перерозподілити персонал та асортимент готової їжі з обіду (13-15) на вечір (18-20), коли реальний попит найвищий.

###Гіпотеза №3
Чи правда, що дрібні покупки (Total < 50) частіше відбуваються в певних категоріях товарів (наприклад, Food and beverages або Health and beauty), ніж великі покупки (Total > 500)?

In [26]:
# Дрібні покупки
small_purchases = df[df['Total'] < 50]

# Великі покупки
large_purchases = df[df['Total'] > 500]

print(f"Дрібних покупок: {len(small_purchases)}")
print(f"Великих покупок: {len(large_purchases)}")

Дрібних покупок: 69
Великих покупок: 227


In [31]:
# Топ категорії для дрібних покупок
print("=== Топ категорії для дрібних покупок (<50) ===")
small_categories = small_purchases['Product line'].value_counts()
print(small_categories)
print(f"\nВідсотки:")
print((small_categories / len(small_purchases) * 100).round(3))

=== Топ категорії для дрібних покупок (<50) ===
Product line
Fashion accessories       16
Health and beauty         13
Sports and travel         13
Electronic accessories    11
Home and lifestyle         8
Food and beverages         8
Name: count, dtype: int64

Відсотки:
Product line
Fashion accessories       23.188
Health and beauty         18.841
Sports and travel         18.841
Electronic accessories    15.942
Home and lifestyle        11.594
Food and beverages        11.594
Name: count, dtype: float64


In [34]:
# Топ категорії для великих покупок
print("\n=== Топ категорії для великих покупок (>500) ===")
large_categories = large_purchases['Product line'].value_counts()
print(large_categories)
print(f"\nВідсотки:")
print((large_categories / len(large_purchases) * 100).round(3))


=== Топ категорії для великих покупок (>500) ===
Product line
Sports and travel         40
Health and beauty         38
Electronic accessories    38
Home and lifestyle        38
Food and beverages        37
Fashion accessories       36
Name: count, dtype: int64

Відсотки:
Product line
Sports and travel         17.621
Health and beauty         16.740
Electronic accessories    16.740
Home and lifestyle        16.740
Food and beverages        16.300
Fashion accessories       15.859
Name: count, dtype: float64


In [37]:
# Створення зведеної таблиці
comparison = pd.DataFrame({
    'Дрібні (<50)': small_purchases['Product line'].value_counts(),
    'Великі (>500)': large_purchases['Product line'].value_counts()
}).fillna(0).astype(int)

# Додавання відсотків
comparison['Дрібні %'] = (comparison['Дрібні (<50)'] / len(small_purchases) * 100).round(3)
comparison['Великі %'] = (comparison['Великі (>500)'] / len(large_purchases) * 100).round(3)

# Сортування за дрібними покупками
comparison = comparison.sort_values('Дрібні (<50)', ascending=False)

print(comparison)

                        Дрібні (<50)  Великі (>500)  Дрібні %  Великі %
Product line                                                           
Fashion accessories               16             36    23.188    15.859
Health and beauty                 13             38    18.841    16.740
Sports and travel                 13             40    18.841    17.621
Electronic accessories            11             38    15.942    16.740
Food and beverages                 8             37    11.594    16.300
Home and lifestyle                 8             38    11.594    16.740


####Аналіз спростував очікування:

#####Великі покупки (>500) не концентруються на дорогих категоріях, а рівномірно розподілені по всіх категоріях (16-17% кожна), що свідчить про комплексні сімейні закупівлі.

#####Натомість дрібні покупки (<50) тяжіють до Fashion accessories (23%) як імпульсні покупки, тоді як Food and beverages найменш популярна для дрібних чеків (11.6%).

#####Бізнесу варто залучати великих покупців через VIP-програми та комплексні знижки, а для дрібних — розміщувати Fashion accessories біля кас для стимулювання імпульсних покупок.